src/lib.rs

```diff
@@ -74,208 +82,377 @@ impl<F: FnOnce()> FnBox for F {
-                        Err(..) => {
-                            // The pool was dropped.
-                            break
-                        }
-                    }
-                }
-            });
-
-            threads.push(ThreadData {
-                _thread_join_handle: thread,
-                pool_sync_rx: pool_sync_rx,
-                thread_sync_tx: thread_sync_tx,
-            });
-        }
-
-        Pool {
-            threads: threads,
-            job_sender: Some(job_sender),
-        }
+    pub fn new(n: isize) -> Pool {
+
+        let thread_done = Arc::new(RwLock::new(()));
+        let (job_tx, job_rx) = channel();
+        let job_rx = Arc::new(Mutex::new(job_rx));
+        adjust_helper(n, &job_rx, &thread_done, &job_tx);
+
+        let p = Pool{
+            job_tx: Mutex::new(job_tx),
+            job_rx: job_rx,
+            count: AtomicIsize::new(n),
+            thread_done: ThreadDone(thread_done),
+        };
+        p
     }
 
-    /// Borrows the pool and allows executing jobs on other
-    /// threads during that scope via the argument of the closure.
+    /// Create a manager for the pool. Pools can have multiple managers at the same time.
     ///
-    /// This method will block until the closure and all its jobs have
-    /// run to completion.
-    pub fn scoped<'pool, 'scope, F, R>(&'pool mut self, f: F) -> R
-        where F: FnOnce(&Scope<'pool, 'scope>) -> R
-    {
-        let scope = Scope {
+    /// Creating a manager from a pool requires acquiring a lock, so cloning an existing manager
+    /// for a pool is usually cheaper.
+    pub fn manager<'pool>(&'pool self) -> Manager<&'pool Pool> {
+        Manager{
+            jobs: 0,
+            job_tx: self.job_tx.lock().unwrap().clone(),
+            results: channel(),
             pool: self,
-            _marker: PhantomData,
-        };
-        f(&scope)
+        }
     }
 
-    /// Returns the number of threads inside this pool.
-    pub fn thread_count(&self) -> u32 {
-        self.threads.len() as u32
+    /// Gets the number of threads
+    pub fn threads(&self) -> isize {
+        self.count.load(Ordering::Relaxed)
     }
 }
 
-/////////////////////////////////////////////////////////////////////////////
 
-/// Handle to the scope during which the threadpool is borrowed.
-pub struct Scope<'pool, 'scope> {
-    pool: &'pool mut Pool,
-    // The 'scope needs to be invariant... it seems?
-    _marker: PhantomData<::std::cell::Cell<&'scope mut ()>>,
+// .88b  d88.  .d8b.  d8b   db  .d8b.   d888b  d88888b d8888b.
+// 88'YbdP`88 d8' `8b 888o  88 d8' `8b 88' Y8b 88'     88  `8D
+// 88  88  88 88ooo88 88V8o 88 88ooo88 88      88ooooo 88oobY'
+// 88  88  88 88~~~88 88 V8o88 88~~~88 88  ooo 88~~~~~ 88`8b
+// 88  88  88 88   88 88  V888 88   88 88. ~8~ 88.     88 `88.
+// YP  YP  YP YP   YP VP   V8P YP   YP  Y888P  Y88888P 88   YD
+
+/// A manager for controlling a threadpool.
+pub struct Manager<P: Borrow<Pool>> {
```

Comment:
It enables Manager<&Pool> if you only want to borrow the pool for a limited scopeManager<Pool> if you want a single manager to own the pool, Manager<Arc<Pool>>